pageserver/src/virtual_file/owned_buffers_io/write/flush.rs

```diff
@@ -193,6 +196,14 @@ where
         handle.join_handle.await.unwrap()
     }
 
+    pub fn shutdown_no_flush(mut self) {
+        let handle = self
+            .inner
+            .take()
+            .expect("must not use after we returned an error");
+        handle.join_handle.abort();
+    }
```

Comment:
So, this would be one of the first instances where we actually exercise the cancel-safety of tokio-epoll-uring.

Reply:
Actually this isn't safe. 

Our caller (`impl Drop for DeltaLayerWriter`) will delete the file from the file system.

But the flush task uses VirtualFile.

With VirtualFile, it can theoretically happen that the file descriptor cache has a miss and we need to re-open the file.

That reopen will either fail (good) or it could open a new inode at that path (bad).

Reply:
if instead of random suffix we would use strongly monotonically increasing numbers for the delta|image layer writer temp files, then I guess there would be no risk of the (bad) case above.

https://github.com/neondatabase/neon/blob/e05b5d6ae4a7dfd58ae0cfe62403a7ae340b3356/pageserver/src/tenant/storage_layer/delta_layer.rs#L302-L316

Reply:
> With VirtualFile, it can theoretically happen that the file descriptor cache has a miss and we need to re-open the file.

In the case you mentioned, since we set `create` options to `false` for the `reopen_options`, I think we are good? The flush task will fail to reopen the file the file has been removed. 

https://github.com/neondatabase/neon/blob/7fa986bc923b6af3c77e888be96123f7ba068313/pageserver/src/virtual_file.rs#L612-L620.

Reply:
> The flush task will fail to reopen the file the file has been removed.

Yeah, the flush task will exit at some later point after the drop (not under our control).

But what if in the meantime the path is occupied by a new file.

There is a schedule where the flush task _could_ would then write into that new file, and that would be corruption:

```
flush task: returns from recv.await with some flush work
T1: drops FlushHandle
T1: std::fs::remove()'s the file from the filesystem
T1: creates a new file with same name in the filesystem (same layer file name and same `rand_string`)
flush task: tries to do the flush work through `VirtualFile::write_...`
flush_task: the VirtualFile::write_ needs to do reopen
flush_task: the VirtualFile::write_ writes into the new file
```

I think if we replace the `rand_string` with a monotonic counter, like we do with EphemeralFile names, we eliminate this case.

Basically it was always the implicit requirement of VirtualFile that

* either you make sure all the VirtualFile instances referencing a path are dropped before you delete it (now-unused `VirtualFile::remove`. API guarantees it by consuming `self`)
* or you disambiguate through the file name, then it's safe to delete a path for which there are still VirtualFile objects in existence, these will simply encounter errors and those need to be dealt with.

The use of `Arc<VirtualFile>` kind of puts ourselves into a corner where we have to do both.

Reply:
Actually we can just spawn the shutdown into a background tokio task.

You only need to keep the Timeline gate open once more.

```rust
fn drop(&mut self) {
   let Some(inner) = self.inner.take() else {
     return;
   };
   tokio::spawn(async move {
     let vfile = inner.blob_writer.shutdown().await;
     // it's ok to use sync io for namespace operations
     // https://github.com/neondatabase/neon/issues/7370
     if let Err(e) = std::fs::remove_file(vfile.path())  {
       e.maybe_fatal_err();
       // nothing we can do about it at this point
       error!(err=%e, path=%vfile.path(), "failed to remove layer writer file");
     }
     // the task keeps the timeline gate open until it's done operating on the timeline dir
     let Inner { gate_guard, .. } = { inner };
     drop(gate_guard);
   })
}
```

Reply:
b1edbf667c61d842db0b7c4076d0ae98433b1128 adds the monotonically increasing file suffix.
38f1150ae29bda4952c52e0bdae9346e21e78a69 use the background task for dropping BlobWriter.